### PR TITLE
[WIP] SameDiff losses/differentiation overhaul

### PR DESCRIPTION
**Pull request by AlexDBlack**
Allows users to specify which variables represent the losses. Also differentiates only the subset of the graph that is actually needed (more efficient). Consequently, it's now possible to train only the subset of the graph that the loss function depends on.

Fixes: https://github.com/deeplearning4j/deeplearning4j/issues/6876
Fixes: https://github.com/deeplearning4j/deeplearning4j/issues/7424 (though not in the proposed way)

Some other major changes:
* VARIABLE type SDVariables can now only be floating point. Variable type means "trainable" but only FP are trainable by backprop, by definition.

### CODE REVIEW

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 1]
```diff
@@ -1427,6 +1429,52 @@ public long numElements() {
...
 
+    /**
+     * Get the names of variables (if any) that have been marked as loss variables to be minimized.<br>
+     * Variables can be marked as loss variables in a few different ways:<br>
+     * (a) Losses are automatically added when creating loss functions via {@link #sd()}<br>
+     * (b) Via {@link #setLossVariables(String...)}, @link #addLossVariable(String)} or {@link SDVariable#markAsLoss()}<br>
+     * (c) Via {@link TrainingConfig#setLossVariables(List)}<br>
+     */
+    public List<String> getLossVariables(){
+        return this.lossVariables;
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: do we want to return a mutable version of the list here?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 2]
```diff
@@ -3122,70 +3356,150 @@ public void createGradFunction() {
...
-                // available before we try to differentiate the op
-                Queue<DifferentialFunction> availableForDiff = new LinkedList<>();
-                Set<String> seenOps = new HashSet<>();
-                //start with scalar backprop
-                INDArray initGradArr;
-                try(MemoryWorkspace ws = Nd4j.getMemoryManager().scopeOutOfWorkspaces()) {
-                    initGradArr = Nd4j.scalar(1.0);
+
+                //----- Step 1: Determine FP variables connected to loss -----
+                // Find all FP variables that are connected to loss by an FP32 path
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: Should this actually be only an fp32 path?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 3]
```diff
@@ -3198,48 +3512,120 @@ public void createGradFunction() {
...
+                            //If no gradient exists at this point, 3 possibilities:
+                            // (a) we have a bug
+                            // (b) output of this op isn't used in calculating the loss
+                            // (c) output isn't a FP type
+                            //In the FP case, we should create a zero variable to backprop, because we can't perform backprop
+                            // for this op otherwise...
+                            if(!v.dataType().isFPType()){
+                                grads.add(null);
+                            } else {
+                                SDVariable gTemp = sameDiff.zerosLike(v);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: Should this even happen? Calculating with a zero gradient, should basically be "don't do any change at all" and therefore we shouldn't be performing backprop at all in that case?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 4]
```diff
@@ -3198,48 +3512,120 @@ public void createGradFunction() {
...
+                            }
+                            if(anyInputsRequired){
+                                if(!differentiatedOps.contains(op.getName())){
+                                    isRequiredOp = true;
+                                }
+                            }
+                        }
+
+                        if(!isRequiredOp){
+                            System.out.println("Skipped as not required: " + opName);
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: Should probably be a `log.info` or `log.trace`?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 5]
```diff
@@ -3198,48 +3512,120 @@ public void createGradFunction() {
...
+                                }
+                                //If in't not in the minimal subgraph, loss doesn't depend on it, so we don't care about it
+                            }
                         }
 
-                        if(allAvailable && !seenOps.contains(o.getOp().getOwnName())) {
+                        if(allAvailable && !availableForDiff.contains(o.getOp())){
                             availableForDiff.add(o.getOp());
-                            seenOps.add(o.getOp().getOwnName());
+                            System.out.println("Marked available for diff: " + o.getOp().getOwnName());
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: again, use logger?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java [hunk 6]
```diff
@@ -3198,48 +3512,120 @@ public void createGradFunction() {
...
+                        } else {
+                            if(!availableForDiff.contains(o.getOp())) {
+                                System.out.println("Not all inputs available: " + o.getName());
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: again, use logger?

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/TrainingConfig.java [hunk 7]
```diff
@@ -348,17 +357,23 @@ public Builder skipBuilderValidation(boolean skip){
...
             if(!skipValidation) {
                 Preconditions.checkState(updater != null, "Updater (optimizer) must not be null. Use updater(IUpdater) to set an updater");
                 Preconditions.checkState(dataSetFeatureMapping != null, "No DataSet feature mapping has been provided. A " +
-                        "mapping between DataSet array positions and variables/placeholders must be provided - use  dateSetFeatureMapping(...) to set this");
-                Preconditions.checkState(dataSetLabelMapping != null, "No DataSet label mapping has been provided. A " +
-                        "mapping between DataSet array positions and variables/placeholders must be provided - use  dateSetLabelMapping(...) to set this");
+                        "mapping between DataSet array positions and variables/placeholders must be provided - use dateSetFeatureMapping(...) to set this");
+                Preconditions.checkState(markLabelsUnused || dataSetLabelMapping != null, "No DataSet label mapping has been provided. A " +
+                        "mapping between DataSet array positions and variables/placeholders must be provided - use dateSetLabelMapping(...) to set this," +
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/TrainingConfig.java]: typo: dateSetFeatureMapping and dateSetLabelMapping should be dataSetFeatureMapping and dataSetLabelMapping

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java (hunk 2 above)
Comment (AlexDBlack) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: No, any FP is fine. Good catch.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java (hunk 1 above)
Comment (AlexDBlack) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: Probably not.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java (hunk 3 above)
Comment (AlexDBlack) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: This is a little unintuitive, but it's definitely intended (and required).
Consider the following graph:
```
input -> split -> (A,B)
B -> lossFn
```
Here, A is unused (in that it doesn't contribute to the loss function). But to do split op backprop, we need gradient variables/arrays for both outputs (A and B).
We know the shape and type of dL/dA must be exactly the same as A; we also know that the loss function doesn't depend on A. Hence, dL/dA == zerosLike(A)

Comment (treo) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: got it. Maybe add the explanation to the comment on top?

Comment (AlexDBlack) [reply nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java]: Done.
